server/datastore/mysql/policies.go

```diff
@@ -134,30 +135,34 @@ func (ds *Datastore) SavePolicy(ctx context.Context, p *fleet.Policy, shouldRemo
 		return ctxerr.Wrap(ctx, notFound("Policy").WithID(p.ID))
 	}
 
-	return ds.cleanupPolicy(ctx, p.ID, p.Platform, shouldRemoveAllPolicyMemberships, removePolicyStats)
+	return cleanupPolicy(ctx, ds.writer(ctx), p.ID, p.Platform, shouldRemoveAllPolicyMemberships, removePolicyStats, ds.logger)
 }
 
-func (ds *Datastore) cleanupPolicy(
-	ctx context.Context, policyID uint, policyPlatform string, shouldRemoveAllPolicyMemberships bool, removePolicyStats bool,
+func cleanupPolicy(
+	ctx context.Context, extContext sqlx.ExtContext, policyID uint, policyPlatform string, shouldRemoveAllPolicyMemberships bool,
+	removePolicyStats bool, logger kitlog.Logger,
 ) error {
 	var err error
 	if shouldRemoveAllPolicyMemberships {
-		err = ds.cleanupPolicyMembershipForPolicy(ctx, policyID)
+		err = cleanupPolicyMembershipForPolicy(ctx, extContext, policyID)
 	} else {
-		err = cleanupPolicyMembershipOnPolicyUpdate(ctx, ds.writer(ctx), policyID, policyPlatform)
+		err = cleanupPolicyMembershipOnPolicyUpdate(ctx, extContext, policyID, policyPlatform)
 	}
 	if err != nil {
 		return err
 	}
 	if removePolicyStats {
 		// delete all policy stats for the policy
-		// wrapping in a retry to avoid deadlocks with the cleanups_then_aggregation cron job
-		err = ds.withRetryTxx(
-			ctx, func(tx sqlx.ExtContext) error {
-				_, err := tx.ExecContext(ctx, `DELETE FROM policy_stats WHERE policy_id = ?`, policyID)
-				return err
-			},
-		)
+		fn := func(tx sqlx.ExtContext) error {
+			_, err := tx.ExecContext(ctx, `DELETE FROM policy_stats WHERE policy_id = ?`, policyID)
+			return err
+		}
+		if _, isDB := extContext.(*sqlx.DB); isDB {
+			// wrapping in a retry to avoid deadlocks with the cleanups_then_aggregation cron job
+			err = withRetryTxx(ctx, extContext.(*sqlx.DB), fn, logger)
```

Comment:
Calling withRetryTxx within withRetryTxx? (ApplyPolicySpecs -> withRetryTxx(cleanupPolicy))

Reply:
Yes, this is what I just fixed.

`withRetryTxx` is only called for `*sqlx.DB` input. ApplyPolicySpecs passes in a `tx`, so `withRetryTxx` doesn't get called twice.

Reply:
Gotcha, I missed this on my last hour of Friday :)